### ISSUE TITLE
Feature Request: Scriptable Config

### ISSUE THREAD

**Issue by tcf909** [ncw/rclone#1010]
Feature Request:

It would be really nice to be able to do scripted installs / configs for use in containers.

Example:

On container init:
```
rclone config --interactive httpd --httpdPort 12345 --httpdRedirectOnDone {{IP}}:54321 --confoption NAME=ACD --confoption TYPE=ACD --confoption CLIENT_ID=value --confoption CLIENT_SECRET=value --confoption WHATEVER=value && \
rclone config --interactive httpd --httpdPort 54321 --confoption TYPE=crypt --confoption REMOTE=ACD:Vault --confoption FILENAME_ENCRYPTION=standard && \
mkdir /mnt/ACD_CRYPT && rclone mount ACD_CRYPT: /mnt/ACD_CRYPT &
```

Basically rather than provide CLI input, have the command launch the built in HTTP server that already exists, and wait for the needed input by the command. The individual commands can take the options provided and present back the HTTP server which options they still need filled in (dynamic input). Once submitted, the redirect kicks in and sends it to the next port / command (or something like that).

I'm going to start working on a small node.js service to manually build the rclone.conf in this manner, but I'd much rather it be handled natively. Right now I see now way to do remotely attended installs without the command line input.

Thoughts?

**Comment by ncw** [comment]
I recently merged something which will allow you to configure rclone through environment variables - see #616 - does that help?

I'm not sure how you would script the oauth login - how do you see that working?

**Comment by tcf909** [comment]
It's definitely a step in the right direction.

The biggest hurdle I ran in to last night (high level) was generating the .rclone.conf. Breaking down the detail of the hurdle, the two primary issues are: 1) Inputing the obscured password and 2) generating the oauth token.

In general, if we want to consider rclone (binary) to be usable in a hands off way, it needs to require a no input workflow. I saw that there are auto_config options in the internal functions, but the primary workflow (rclone config) definitely requires input (even to use the environmental variables) from what I understand.

For the Oauth portion: I think the path of least resistance would be to have a flag that would trigger the http server with the right id and name and await web input).

Something like:

rclone writeconfig --name NAME --TYPE ACD --OTHEROPT OTHERVALUE --auth http --httpPort 12345 --httpRedirect http://www.whatever.com --httpBind 0.0.0.0

This could write the initial config and then launch the http server that is there with the right variables (and maybe a timeout by default).

I think all the parts are there, just needs a new workflow path.

**Comment by tamama** [comment]
As it is 2019 today, may I ask if there are any best practice for non-interactive config (or at least semi-interactive - only the oauth part) regarding rclone config?

**Comment by ncw** [comment]
@tamama 

check out the `rclone config` sub commands - you can use these to create config values.

```
$ rclone config -h
Enter an interactive configuration session where you can setup new
remotes and manage existing ones. You may also set or remove a
password to protect your configuration.

Usage:
  rclone config [flags]
  rclone config [command]

Available Commands:
  create      Create a new remote with name, type and options.
  delete      Delete an existing remote <name>.
  dump        Dump the config file as JSON.
  edit        Enter an interactive configuration session.
  file        Show path of configuration file in use.
  password    Update password in an existing remote.
  providers   List in JSON format all the providers and options.
  show        Print (decrypted) config file, or the config for a single remote.
  update      Update options in an existing remote.

Flags:
  -h, --help   help for config

Use "rclone [command] --help" for more information about a command.
Use "rclone help flags" for to see the global flags.
Use "rclone help backends" for a list of supported services.
```

Note that you can use rclone without a config file - every config file value can be passed in as a flag or as an environment variable.

**Comment by tamama** [comment]
@ncw May I know the flag to disable auto-config? (My server is a headless one...)

Here is a simple log:
--------------------
[root@london-node-230 ~]# rclone config create gdrive drive --drive-scope="drive" --auto-confirm=false
Remote config
If your browser doesn't open automatically go to the following link: http://127.0.0.1:53682/auth
Log in and authorize rclone for access
Waiting for code...

**Comment by tamama** [comment]
(cont'd) What I really want is:

Remote config
Use auto config?
 * Say Y if not sure
 * Say N if you are working on a remote or headless machine or Y didn't work
y) Yes
n) No
y/n> n
If your browser doesn't open automatically go to the following link: https://accounts.google.com/o/oauth2/auth?access_type=offline&client_id=202264815644.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&response_type=code&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive&state=301258b83e6f52f5ab58dc9043719b00
Log in and authorize rclone for access

**Comment by tamama** [comment]
@ncw Would this be possible? Thanks.

**Comment by ncw** [comment]
@tamama I've had a go at fixing this.

Please try

https://beta.rclone.org/branch/v1.45-095-g40179518-fix-1010-script-config-beta/ (uploaded in 15-30 mins)

You'll need to use this (note the parameters are `key` `value` pairs from the config file).

    rclone config create gdrive drive scope drive config_is_local false

**Comment by tamama** [comment]
@ncw This works perfectly! (The rclone binary that I used is from rclone-v1.45-095-g40179518-fix-1010-script-config-beta-linux-amd64.zip)

Please kindly tell me when this patch will be available in the official beta branch, so that 
curl https://rclone.org/install.sh | bash -s beta 

would work.

Thanks!

**Comment by tamama** [comment]
@ncw I think that your rclone project would be a killer when integrated with configuration management tools, such as Puppet or Ansible. Truly unbeatable.

**Comment by ncw** [comment]
> @ncw This works perfectly! (The rclone binary that I used is from rclone-v1.45-095-g40179518-fix-1010-script-config-beta-linux-amd64.zip)
> 
> Please kindly tell me when this patch will be available in the official beta branch, so that
> curl https://rclone.org/install.sh | bash -s beta

Great - thanks for testing :-)

I've merged this to master now and it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released with v1.46



> @ncw I think that your rclone project would be a killer when integrated with configuration management tools, such as Puppet or Ansible. Truly unbeatable. Awesomely done!

Thanks for the praise!

There is an ansible integration here: https://github.com/stefangweichinger/ansible-rclone

**Comment by tamama** [comment]
Awesome. Thanks!
